### PR TITLE
[satel] Command and events refactoring, communication improvements

### PR DESCRIPTION
**Pull request by druciak**
This PR contains following changes:
- internal binding event system has been refactored: now each event has its own handler method; this way there is no need to check type of incoming event in handlers;
- basic response validation has been moved to base class `SatelCommandBase`;
- `handleResponse` has been splitted to avoid necessity of calling super class `handleResponse`;
- workaround for https://community.openhab.org/t/satel-zones-not-work/91620/15 in `SatelModule` has been added - it ignores responses that don't match sent command;

Signed-off-by: Krzysztof Goworek <krzysztof.goworek@gmail.com>

### CODE REVIEW

File: bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/event/EventDispatcher.java [hunk 1]
```diff
@@ -56,10 +67,19 @@ public void removeEventListener(SatelEventListener eventListener) {
      * @param event the event to distribute.
      */
     public void dispatchEvent(SatelEvent event) {
-        logger.debug("Distributing event: {}", event);
-        for (SatelEventListener listener : eventListeners) {
-            logger.trace("Distributing to {}", listener);
-            listener.incomingEvent(event);
+        final Method m = eventHandlers.get(event.getClass());
+        if (m == null) {
+            logger.error("Missing event handler for event {}. Event discarded.", event.getClass().getName());
```
Comment (J-N-K) [inline bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/event/EventDispatcher.java]:
```suggestion
            logger.warn("Missing event handler for event {}. Event discarded.", event.getClass().getName());
```

File: bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/event/EventDispatcher.java [hunk 2]
```diff
@@ -56,10 +67,19 @@ public void removeEventListener(SatelEventListener eventListener) {
...
+        } else {
+            logger.debug("Distributing event: {}", event);
+            for (SatelEventListener listener : eventListeners) {
+                logger.trace("Distributing to {}", listener);
+                try {
+                    m.invoke(listener, event);
+                } catch (ReflectiveOperationException e) {
+                    logger.error("Unable to distribute {} to {}", event.getClass().getName(), listener, e);
```
Comment (J-N-K) [inline bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/event/EventDispatcher.java]:
```suggestion
                    logger.warn("Unable to distribute {} to {}", event.getClass().getName(), listener, e);
```

File: bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/event/EventDispatcher.java [hunk 3]
```diff
@@ -56,10 +67,19 @@ public void removeEventListener(SatelEventListener eventListener) {
...
-        logger.debug("Distributing event: {}", event);
-        for (SatelEventListener listener : eventListeners) {
-            logger.trace("Distributing to {}", listener);
-            listener.incomingEvent(event);
+        final Method m = eventHandlers.get(event.getClass());
+        if (m == null) {
+            logger.error("Missing event handler for event {}. Event discarded.", event.getClass().getName());
+        } else {
+            logger.debug("Distributing event: {}", event);
+            for (SatelEventListener listener : eventListeners) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/event/EventDispatcher.java]:
```suggestion
           eventListeners.forEach(listener -> {
```

File: bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/command/SatelCommandBase.java [hunk 4]
```diff
@@ -115,6 +125,14 @@ protected SatelMessage getResponse() {
      */
     protected abstract boolean isResponseValid(SatelMessage response);
 
+    /**
+     * Overriden in subclasses allows to execute action specific to given command (i.e. dispatch an event).
+     *
+     * @param eventDispatcher event dispatcher
+     */
+    protected void handleResponseInternal(final EventDispatcher eventDispatcher) {
```
[the remark this answers is not in the document]
Comment (druciak) [reply bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/command/SatelCommandBase.java]: This is on purpose. Not all subclasses need to override this method, so it is empty by default.

[the remark this answers is not in the document]
Comment (J-N-K) [reply bundles/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/command/SatelCommandBase.java]: I guess that‘s a personal style question. I prefer „deliberately implement to do nothing in subclass“ over „accidentally forget to implement doing something“. But ok with me.

**Comment by J-N-K** [comment]
Please ping me if I forget to merge once Jenkins finished.

**Comment by druciak** [comment]
Sure, I will. Thank you!

**Comment by druciak** [comment]
Ready to merge.

**Pull request merged**